[ci] [R-package] clang16 CI jobs failing: "error: templates must have C++ linkage"

## Description

`(clang16, R-devel)` CI jobs have been failing here for the last few days, with this error in the installation of `{fs}`

```text
In file included from error.cc:1:
In file included from ./error.h:5:
In file included from /opt/R/devel/lib/R/include/Rinternals.h:47:
In file included from /opt/R/devel/lib/R/include/R_ext/Error.h:30:
In file included from /opt/R/devel/lib/R/include/R_ext/RS.h:30:
/usr/lib/llvm-16/bin/../include/c++/v1/cstring:117:1: error: templates must have C++ linkage
template <class _Tp>
^~~~~~~~~~~~~~~~~~~~
/opt/R/devel/lib/R/include/Rinternals.h:37:1: note: extern "C" language linkage specification begins here
extern "C" {
^
1 error generated.
make[1]: *** [/opt/R/devel/lib/R/etc/Makeconf:202: error.o] Error 1
make[1]: Leaving directory '/tmp/RtmpZZyG7J/R.INSTALLb894784240/fs/src'
ERROR: compilation failed for package ‘fs’
```

That looks like an incompatibility between R's own headers and `clang-16`, not something LightGBM needs to fix.

## Reproducible example

This is happening in all CI builds.

Example from #6600: ([build link](https://github.com/microsoft/LightGBM/actions/runs/10332413944/job/28738187064?pr=6600))

## Environment info

N/A

## Additional Comments

`{fs}` is passing this check currently on CRAN: https://cran.r-project.org/web/checks/check_results_fs.html

CRAN is still running the clang16 check: https://cran.r-project.org/web/checks/check_issue_kinds.html

--- a/.github/workflows/r_package.yml
+++ b/.github/workflows/r_package.yml
@@ -268,9 +268,7 @@ jobs:
         #   * CRAN "additional checks": https://cran.r-project.org/web/checks/check_issue_kinds.html
         #   * images: https://r-hub.github.io/containers/containers.html
         image:
-          # clang16 should be re-enabled once it's fixed upstream
-          # ref: https://github.com/microsoft/LightGBM/issues/6607
-          #- clang16
+          - clang16
           - clang17
           - clang18
           - clang19

